### ISSUE TITLE
Lack of standard for deriving connect-Line style from connector icon

### ISSUE THREAD

**Issue by henrikt-ma** [modelica/ModelicaSpecification#3380]
I think that the practice of deriving the styling for a new `connect` `Line`-annotation based on connector graphics is more or less a de facto standard.  It should be made a proper standard.

(Once there is a standard, the `Modelica.UsersGuide.Conventions.Icons` should also be updated, but that's an issue for the [ModelicaStandardLibrary](https://github.com/modelica/ModelicaStandardLibrary) project.)

**Comment by henrikt-ma** [comment]
As a starting point for the discussion, my understanding that all style attributes should be taken from the first graphical primitive of the connector class' `Icon`-annotation.  Further, these are the attributes that would be transferred to the `connect` `Line`-annotation:
- `lineColor` (mapped to `color` of the `Line`)
- `pattern` (why on earth does `FilledShape` call it just `pattern` and not `linePattern`?)
- `lineThickness` (mapped to `thickness` of the `Line`)
- `smooth`

In particular, none of the properties from `GraphicItem` are transferred (allowing the primitive to have `visible = false` if it would disturb the look of the icon).  Further, a `Line` would not be usable, as it has the attributes `color` and `thickness` rather than `lineColor` and `lineThickness`.

Wouldn't it make a lot of sense to be able to use a `Line` instead of a filled shape?

**Comment by henrikt-ma** [comment]
I suppose that all compatible connectors are expected to have consistent attributes affecting connection styling, so that it doesn't matter whether attributes are taken from `a` or `b` in `connect(a, b)`.  Maybe this should be mentioned as a non-normative recommendation in the specification, in addition to clearly recommending it in `Modelica.UsersGuide.Conventions.Icons`?

**Comment by maltelenz** [comment]
> I suppose that all compatible connectors are expected to have consistent attributes affecting connection styling, so that it doesn't matter whether attributes are taken from `a` or `b` in `connect(a, b)`. Maybe this should be mentioned as a non-normative recommendation in the specification, in addition to clearly recommending it in `Modelica.UsersGuide.Conventions.Icons`?

I don't think this is the case when using expandable connectors for buses, see for example [Modelica.Mechanics.MultiBody.Examples.Systems.RobotR3.Utilities.AxisType2](https://doc.modelica.org/Modelica%204.0.0/Resources/helpWSM/Modelica/Modelica.Mechanics.MultiBody.Examples.Systems.RobotR3.Utilities.AxisType2.html), where the two ends of connections to the sensors differ.

**Comment by henrikt-ma** [comment]
Right.  Could it be as simple as saying that the connection line style should never be inferred from the icon of an expandable connector, so that one doesn't need to worry about styling of the first graphical primitive of an expandable connector icon?

**Comment by maltelenz** [comment]
Or that the styling is inferred from the starting/first side of the connection. This allows the user to choose the styling by starting the connection at the corresponding end.

**Comment by HansOlsson** [comment]
> Right. Could it be as simple as saying that the connection line style should never be inferred from the icon of an expandable connector, so that one doesn't need to worry about styling of the first graphical primitive of an expandable connector icon?

That doesn't work when connecting between two expandable connectors, as is also seen in Modelica.Mechanics.MultiBody.Examples.Systems.RobotR3.Utilities.AxisType2 (the two yellow lines).

An alternative would be to prefer another connector, or care whether it is a sub-part or not.

> Or that the styling is inferred from the starting/first side of the connection. This allows the user to choose the styling by starting the connection at the corresponding end.

Could possibly do, but I'm not sure if we need to have those details specified.

**Comment by henrikt-ma** [comment]
> > Or that the styling is inferred from the starting/first side of the connection. This allows the user to choose the styling by starting the connection at the corresponding end.
> 
> Could possibly do, but I'm not sure if we need to have those details specified.

Right.  It would have been different if the `connect`-equation didn't have its own `Line` styling, so that the line would dynamically change style when the connector icon changes.  What is needed now is just a convention that gives library developers a tool-independent way of suggesting the `Line` style for new connections.  Perhaps the non-normative recommendation should simply be to make an active choice of `Line` style hint for every connector icon?

**Comment by DagBruck** [comment]
After looking at the Dymola code, let me add a few points to Henrik's description.
* Dymola also sets line color, line pattern and thickness attributes. The others are reset to defaults, including smooth.
* We copy the attributes from the **first** enabled rectangle or polygon primitive, no others. Not from circles, for example. I'm sure there was a discussion behind that decision, but I have now forgotten the motivation.
* For some obvious cases we do not copy an attribute: line pattern "none" and line thickness <= 0. This means that you can in fact set the color of the connection to a different color from the connector, by using one color as fill color and and different color for line color (with pattern "none").

**Comment by henrikt-ma** [comment]
> * We copy the attributes from the **first** enabled rectangle or polygon primitive, no others. Not from circles, for example. I'm sure there was a discussion behind that decision, but I have now forgotten the motivation.

By _enabled_, do you mean the absence of `visible = false`?  (Would sound like an unnecessary restriction to me, especially in view of allowing the trick to set `LinePattern.None` and `FillPattern.None` – do you remember any motivation behind this restriction?)

**Comment by DagBruck** [comment]
> > * We copy the attributes from the **first** enabled rectangle or polygon primitive, no others. Not from circles, for example. I'm sure there was a discussion behind that decision, but I have now forgotten the motivation.
> 
> By _enabled_, do you mean the absence of `visible = false`? (Would sound like an unnecessary restriction to me, especially in view of allowing the trick to set `LinePattern.None` and `FillPattern.None` – do you remember any motivation behind this restriction?)

Yes, that's what I mean by enabled (or rather what the comment in the code said). No, I don't remember the motivation anymore. Maybe that `visible=false` should mean not only out-of-sight but also out-of-mind? It may be an unnecessary restriction, but it's what Dymola does today.

**Comment by HansOlsson** [comment]
> Regarding `smooth`, this is something we (System Modeler) would very much like to be considered part of the styling.

To me a smooth outline of an element in a connector (which is quite small) is quite different from the smooth-ness of the connection-line.

**Comment by henrikt-ma** [comment]
Tool vendors that are interested in providing input to this discussion are welcome to show the result of making a connection between the two components in this model:
```
model DefaultConnectLineStyle
  connector C
    annotation(Diagram(coordinateSystem(extent = {{-100, -100}, {100, 100}}, preserveAspectRatio = true, initialScale = 0.1, grid = {10, 10})), Icon(coordinateSystem(extent = {{-10, -10}, {10, 10}}, preserveAspectRatio = true, initialScale = 1, grid = {1, 1}), graphics = {Polygon(lineColor = {255, 0, 255}, fillColor = {0, 255, 0}, pattern = LinePattern.Dot, fillPattern = FillPattern.CrossDiag, lineThickness = 3, smooth = Smooth.Bezier, origin = {2.858, -0.155}, points = {{5.756, -3.562}, {6.356, 1.402}, {4.094, 7.708}, {-6.828, 9.02}, {-2.818, 0.481}, {-8.736, -6.865}, {2.176, -8.184}})}));
  end C;

  model M
    C c annotation(Placement(visible = true, transformation(origin = {102.506, 0.283}, extent = {{-10, -10}, {10, 10}}, rotation = 0), iconTransformation(origin = {10, -0}, extent = {{-2.5, -2.5}, {2.5, 2.5}}, rotation = 0)));
    annotation(Diagram(coordinateSystem(extent = {{-100, -100}, {100, 100}}, preserveAspectRatio = true, initialScale = 0.1, grid = {10, 10})), Icon(coordinateSystem(extent = {{-10, -10}, {10, 10}}, preserveAspectRatio = true, initialScale = 1, grid = {1, 1}), graphics = {Rectangle(visible = true, fillColor = {255, 255, 255}, extent = {{-10, -10}, {10, 10}})}));
  end M;

  M m1 annotation(Placement(visible = true, transformation(origin = {30, 30}, extent = {{-10, -10}, {10, 10}}, rotation = 0)));
  M m2 annotation(Placement(visible = true, transformation(origin = {70, 70}, extent = {{-10, -10}, {10, 10}}, rotation = -450)));
  annotation(Diagram(coordinateSystem(extent = {{10, 10}, {90, 90}}, preserveAspectRatio = true, initialScale = 0.1, grid = {5, 5})), Icon(coordinateSystem(extent = {{-100, -100}, {100, 100}}, preserveAspectRatio = true, initialScale = 0.1, grid = {10, 10}), graphics = {Rectangle(visible = true, lineColor = {0, 114, 195}, fillColor = {255, 255, 255}, extent = {{-100, -100}, {100, 100}}, radius = 25), Text(visible = true, textColor = {64, 64, 64}, extent = {{-150, 110}, {150, 150}}, textString = "%name")}));
end DefaultConnectLineStyle;
```

For convenience, the `Polygon` of the connector icon is repeated here:
```
Polygon(
  lineColor = {255, 0, 255},
  fillColor = {0, 255, 0},
  pattern = LinePattern.Dot,
  fillPattern = FillPattern.CrossDiag,
  lineThickness = 3,
  smooth = Smooth.Bezier,
  origin = …,
  points = …
)
```

This is what the diagram looks like before adding the connection:
<img width="300" alt="before-connect" src="https://github.com/modelica/ModelicaSpecification/assets/25294143/88bd427b-2c63-4add-b896-a27da8a26928">

**Comment by henrikt-ma** [comment]
At the moment, System Modeler only picks up the `lineColor`:
```
  connect(m1.c, m2.c) annotation(Line(visible = true, origin = {60, 40}, points = {{-20, -10}, {10, -10}, {10, 20}}, color = {255, 0, 255}));
```

**Comment by henrikt-ma** [comment]
In a development build of System Modeler, the resulting line is picking up all the style attributes mentioned above:
```
  connect(m1.c, m2.c) annotation(Line(visible = true, origin = {60, 40}, points = {{-20, -10}, {10, -10}, {10, 20}}, color = {255, 0, 255}, pattern = LinePattern.Dot, thickness = 3, smooth = Smooth.Bezier));
```
That we set `visible` is a tool-specific behavior that should be disregarded for purposes of this investigation.

**Comment by DagBruck** [comment]
Dymola 2023x Refresh 1 (!), the last official release gives you this connection.

```
connect(m1.c, m2.c)
   annotation (Line(
      points={{40,30},{70,30},{70,60}},
      color={255,0,255},
      pattern=LinePattern.Dot,
      thickness=0.5));
```
My development version of Dymola gives:

```
connect(m1.c, m2.c)
   annotation (Line(
      points={{40,30},{70,30},{70,60}},
      color={255,0,255},
      pattern=LinePattern.Dot,
      thickness=0.5,
      smooth=Smooth.Bezier));
```

**Comment by DagBruck** [comment]
Our development versions seem similar @henrikt-ma, except for the thickness attribute. Please enlighten me.

**Comment by henrikt-ma** [comment]
> Our development versions seem similar @henrikt-ma, except for the thickness attribute. Please enlighten me.

That was my bad.  That was the original `lineThickness` of my example, but I later changed it to get more reasonable looks.  I've now corrected the reported behavior for System Modeler, making it consistent with what you report for the development version of Dymola.

Pinging some people who might be interested in representing other tools: @sjoelund, @gkurzbach, @eshmoylova, @MarkusOlssonModelon

**Comment by casella** [comment]
Adding @adeas31 to the loop, he's the one who develops the GUI

**Comment by henrikt-ma** [comment]
> Note that the `lineThickness` is not included since we have a special handling for it.
> 
> > Only set the connection line thickness to 0.5 when both connectors are either expandable or (array && !connectorSizing). Otherwise set it to 0.25 i.e., default.

Alright, so here we have something that needs to be resolved.  I can understand that line thickness is the most controversial attribute to pick from the connector icon since I can see that leaving the control to the tool has clear advantages, such as keeping the same width in pixels regardless of magnification on screen.

Some alternative resolutions come to mind:
- (a) Specify that `lineThickness` should be considered, but update MSL connector icons and style guide so that the first polygon does not specify `lineThickness`.
- (b) Exclude `lineThickness` from the attributes to pick up.
- (c) Pick up `thickness` from the first `Line` with `visible = false` instead.

I find (a) problematic, since it makes the first polygon less useful as a visible part of the connectors icon due to the inability to set `lineThickness` without unwanted consequences.

The problem with (b), of course, is that it gives the library author no control over the connection line thickness.

The reason that (c) requires `visible = false` is to avoid unwanted consequences for existing connector icons.

**Comment by eshmoylova** [comment]
In MapleSim, we pick up the color attribute only. 
```
	connect(M1.c, M2.c) annotation(Line(points={{550.0,-30.0},{550.0,-130.0},{440.0,-130.0}},color={255,0,255},smooth=Smooth.None));
```
From the discussion with the developers, we specifically pick up the color from the connector that you click on first to draw the line. Usually, when the connector type match the colors also match.

**Comment by henrikt-ma** [comment]
We stumbled upon this again.  I was asked to bring forward a request for the use of an annotation to solve this properly:
- Use a `connectLine`-annotation to go side by side with `icon`, for example `connectLine(thickness = 0.5, color = {255, 0, 255})`.
- When a `connectionLine`-annotation is present, no connect-line styling is taken from `icon`.
- When there is no `connectionLine`-annotation, take a small selection of attributes from the first Polygon in the `icon`; the smaller the better, as it avoids conflicting interpretation with the looks of the connector.  At least, exclude `lineThickness` based on the discussion above.

**Comment by adeas31** [comment]
> When there is no connectionLine-annotation, take a small selection of attributes from the first Polygon in the icon; the smaller the better, as it avoids conflicting interpretation with the looks of the connector. At least, exclude lineThickness based on the discussion above.

Shouldn't it be the first graphical primitive instead of Polygon?

**Comment by henrikt-ma** [comment]
> Shouldn't it be the first graphical primitive instead of Polygon?

Considering the conversation above, I think the only alternative which would be relevant in view of what tools to today would be to say _`Polygon` or `Rectangle`_.

**Comment by adeas31** [comment]
Well if we are talking about standardization then we should not consider it in the context of a tool. Lets say you can have components with just ellipses and lines. So the first graphical primitive makes sense or the first one that extends from `FilledShape`.

**Comment by gwr69** [comment]
Just being curious: What about something like `Modelica.Blocks.Interfaces.RealVectorInput`?

**Comment by henrikt-ma** [comment]
> Just being curious: What about something like `Modelica.Blocks.Interfaces.RealVectorInput`?

Apparently Dymola will use some default `connect`-line styling in this case.  System Modeler, on the other hand, is picking up the color also in this case.  Seems like a mess to standardize behavior for the `Ellipse` today, considering that libraries developed with Dymola could be full of ellipses that were not meant to affect `connect`-line styling.

**Comment by DagBruck** [comment]
> We stumbled upon this again. I was asked to bring forward a request for the use of an annotation to solve this properly:

Can you elaborate what you stumbled on, and how hard? I'm not convinced this is a problem we need to solve, or at least that it is of significantly lower importance than many other aspects of Modelica. And it is not free: it has to be implemented, it needs UI to set up (I have no idea how that would be done), it must be documented.

**Comment by HansOlsson** [comment]
> > We stumbled upon this again. I was asked to bring forward a request for the use of an annotation to solve this properly:
> 
> Can you elaborate what you stumbled on, and how hard? I'm not convinced this is a problem we need to solve, or at least that it is of significantly lower importance than many other aspects of Modelica. And it is not free: it has to be implemented, it needs UI to set up (I have no idea how that would be done), it must be documented.

With the current proposal it is also trivial to get any attributes you want - just create an invisible rectangle (with the appropriate attributes) first in the list.

I also think that Dag _underestimates_ the difficulty of having a special attribute, especially without any fallback.

It is not only a matter of updating all tools (both to internally support it and to have UI for it) - but it also adds:
* A burden for all model developers - as they will now have to make a choice
* A large number of updates in all Modelica source code repositories to store the new attribute

Note that when we allowed moving icons differently in icon and diagram layer we made it clear that the default is keeping the old behavior (=that they are the same and thus shouldn't be stored), which avoids most of these issues - but it still adds complexity for the user interface.

**Comment by henrikt-ma** [comment]
> Can you elaborate what you stumbled on, and how hard? I'm not convinced this is a problem we need to solve, or at least that it is of significantly lower importance than many other aspects of Modelica. And it is not free: it has to be implemented, it needs UI to set up (I have no idea how that would be done), it must be documented.

Someone stumbled on the need to get `connect`-lines with `smooth = Smooth.Bezier`, and that there is:
- 1) No standardized way of achieving this.
- 2) An undocumented hack which is arguably quite inconvenient to use through the icon editor, since selecting the item you need to style requires moving all the graphics on top to the side first.

Of course, implementations wouldn't come for free, but it wouldn't be _that_ difficult to have a button that opens up a dialog for specifying the `connect`-line styling.  An ambitious tool could even show a small `connect`-line preview next to the icon when editing the icon.  For documentation in the specification, we're already doing the hard work in this issue conversation, but I see that there's additional need to also have documentation in tools, like with so many other things we do.

It would be possible to implement most of the ambitious plan without a corresponding annotation in the specification, but it's just a mess for the user that editing the `connect` line style in some cases have visible effect on the icon, and sometimes not.

**Comment by henrikt-ma** [comment]
Bumping this, since I ran into this issue again when finding the need to explain the current behavior to an LLM.
